presto-main/src/main/java/com/facebook/presto/execution/QueryLimitMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,178 @@
+            Function<QueryId, QueryContext> queryContextSupplier,
+            ScheduledExecutorService taskManagementExecutor)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.queryContextSupplier = requireNonNull(queryContextSupplier, "queryContextSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerPoolListeners();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+            scheduledFuture = null;
+        }
+
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool, QueryId queryId)
+    {
+        try {
+            QueryContext queryContext = queryContextSupplier.apply(queryId);
+            long queryTotalMemoryUse = getTotalMemoryUse(queryContext);
+            long maxTotalMemory = queryContext.getMaxTotalMemory();
+            if (queryTotalMemoryUse <= maxTotalMemory) {
+                return;
+            }
+
+            if (checkPending.compareAndSet(false, true)) {
+                log.debug("Scheduling check for %s", queryId);
+                scheduleRevoking(queryContext, queryTotalMemoryUse);
+            }
+        }
+        catch (Throwable e) {
+            log.error(e, "Error when acting on memory pool reservation");
+        }
+    }
+
+    private void scheduleRevoking(QueryContext queryContext, long queryTotalMemoryUse)
+    {
+        taskManagementExecutor.execute(() -> {
+            try {
+                revokeMemory(queryContext, queryTotalMemoryUse);
+            }
+            catch (Throwable e) {
+                log.error(e, "Error requesting memory revoking");
+            }
+        });
+    }
+
+    private long getTotalMemoryUse(QueryContext queryContext)
+    {
+        MemoryPool memoryPool = queryContext.getMemoryPool();
+        QueryId queryId = queryContext.getQueryId();
+        long maxTotalMemory = queryContext.getMaxTotalMemory();
+        return memoryPool.getQueryMemoryReservation(queryId) + memoryPool.getQueryRevocableMemoryReservation(queryId);
+    }
+
+    @VisibleForTesting
+    synchronized void revokeMemory(QueryContext queryContext, long queryTotalMemory)
+    {
+        if (checkPending.getAndSet(false)) {
+            // order tasks by decreasing revocable memory so that we revoke from the tasks with highest revocable memory first
+            PriorityQueue<TaskContext> queryTaskContexts = new PriorityQueue<>(Comparator.comparing(TaskContext::getRevocableMemoryReservation).reversed());
```

Comment:
Worth noting that `TaskContext#getRevocableMemoryReservation` is not necessarily going to be stable between calls through the `Comparator` (and will contend for locks on each access). You might consider capturing a snapshot of these values before sorting them into the priority queue or maybe skipping the sort altogether (how often does a task have a non-zero but trivial amount of memory to revoke so as to not be a NOOP but still be a waste of time to revoke out of order?).

Reply:
I kept an ordering so we know what order to expect revoking for testing purposes, but changed to order by taskId which shouldn't require a lock.